include/boost/url/impl/any_allocator.hpp

```diff
@@ -0,0 +1,190 @@
+try_deallocate_aligned(const Allocator& a, void* p, std::size_t n, std::size_t size, std::size_t align)
+{
+    using aligned_t = max_align_t;
+    using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+    aligned_allocator_t(a).deallocate(reinterpret_cast<aligned_t*>(p), n * ceil_div(size, align));
+}
+
+template <std::size_t N, class Allocator>
+typename std::enable_if<N < alignof(max_align_t), void>::type
+try_deallocate_aligned(const Allocator& a, void* p, std::size_t n, std::size_t size, std::size_t align)
+{
+    if (align <= N)
+    {
+        using aligned_t = typename boost::type_with_alignment<N>::type;
+        using aligned_allocator_t = boost::allocator_rebind_t<Allocator, aligned_t>;
+        aligned_allocator_t(a).deallocate(reinterpret_cast<aligned_t*>(p), n * ceil_div(size, align));
+    } else {
+        try_deallocate_aligned<2*N>(a, p, n, size, align);
+    }
+}
+
+template <class Allocator>
+void deallocate_aligned(const Allocator& a, void* p, std::size_t n, std::size_t size, std::size_t align)
+{
+    try_deallocate_aligned<2*alignof(char)>(a, p, n, size, align);
+}
+}
+
+template <class T>
+any_allocator<T>::
+any_allocator() noexcept
+: any_allocator(std::allocator<char>{})
+{
+}
+
+template <class T>
+template <class Allocator>
+any_allocator<T>::
+any_allocator(Allocator const& a) noexcept
+{
+    class impl;
+
+    using char_allocator_type = boost::allocator_rebind_t<Allocator, char>;
+
+    class impl
+            : public detail::any_allocator_base
+    {
+         char_allocator_type a_;
+
+    public:
+        explicit
+        impl(char_allocator_type const& a)
+        : any_allocator_base(detail::get_type_id<char_allocator_type>()),
+          a_(a)
+        {
+        }
+
+        void*
+        allocate(std::size_t n, std::size_t size, std::size_t align) override
+        {
+            if (align == alignof(char))
+                return a_.allocate(n);
+            else
+                return detail::allocate_aligned(a_, n, size, align);
+        }
+
+        void
+        deallocate(void* p, std::size_t n, std::size_t size, std::size_t align)
+        {
+            if (align == alignof(char))
+                a_.deallocate(reinterpret_cast<char*>(p), n);
+            else
+                detail::deallocate_aligned(a_, p, n, size, align);
+        }
+
+        bool
+        is_equal(const any_allocator_base& other) override
+        {
+            const impl* other_impl = static_cast<const impl*>(&other);
+            BOOST_ASSERT(other_impl);
+            return a_ == other_impl->a_;
+        };
+    };
+
+    p_ = std::allocate_shared<impl>(a, char_allocator_type{});
+}
+
+template <class T>
+template <class U>
+any_allocator<T>::
+any_allocator(
+    any_allocator<U> const& other) noexcept
+: p_(other.p_)
+{
+}
+
+template <class T>
+T*
+any_allocator<T>::
+allocate(
+    std::size_t n) const
+{
+    BOOST_ASSERT(p_);
+    return std::allocator<T>().allocate(n);
```

Comment:
huh? This can't be right... we aren't using `p_`??

Reply:
Ooooops... I removed the wrong line when removing the previous `== nullptr` condition.